<a href="https://colab.research.google.com/github/rmansoura/ai-engineer-journey/blob/main/introPytorch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import torch

In [33]:
my_list = [1,2,3,4]
mylti_dim_list = [[1,2,3,4], [5,6,7,8]]
print(my_list)
print(mylti_dim_list)



[1, 2, 3, 4]
[[1, 2, 3, 4], [5, 6, 7, 8]]


In [38]:
import numpy as np
my_array = np.random.rand(3,4)
print(my_array)

[[0.57272909 0.1051838  0.61971858 0.38428644]
 [0.23982703 0.23525556 0.80641933 0.64293024]
 [0.19484654 0.71312655 0.48570469 0.86052963]]


In [48]:
tensor_2d = torch.rand(3,4)
tensor_2d

tensor([[0.7461, 0.7258, 0.7717, 0.2922],
        [0.5406, 0.2661, 0.9481, 0.3287],
        [0.4065, 0.5712, 0.6704, 0.9994]])

In [54]:
tensor_3d = torch.zeros(2,3,4)
tensor_3d

tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])

# Task
My task is to write and execute Python code using PyTorch to demonstrate the creation, manipulation, and fundamental operations of Tensors.

## Introduction to PyTorch and Tensors

### Subtask:
Understand what PyTorch is, its advantages, and the fundamental concept of Tensors. We will cover creating, manipulating, and performing operations on tensors.


**Reasoning**:
The first step in understanding PyTorch and Tensors is to import the necessary library.



In [4]:
import torch
print("PyTorch imported successfully.")

PyTorch imported successfully.


**Reasoning**:
Now that PyTorch is imported, the next step is to create various types of tensors, starting with creating a tensor from a Python list or NumPy array, as specified in instruction 2a.



In [5]:
import numpy as np

# From a Python list
list_data = [[1, 2, 3], [4, 5, 6]]
tensor_from_list = torch.tensor(list_data)
print("Tensor from list:\n", tensor_from_list)

# From a NumPy array
numpy_array_data = np.array([[7, 8], [9, 10]])
tensor_from_numpy = torch.from_numpy(numpy_array_data)
print("\nTensor from NumPy array:\n", tensor_from_numpy)

Tensor from list:
 tensor([[1, 2, 3],
        [4, 5, 6]])

Tensor from NumPy array:
 tensor([[ 7,  8],
        [ 9, 10]])


**Reasoning**:
Following the instructions, the next step is to create tensors using `torch.empty()`, `torch.zeros()`, and `torch.ones()` with specified shapes, as outlined in instruction 2b.



In [6]:
empty_tensor = torch.empty(2, 3)
print("Empty Tensor:\n", empty_tensor)

zeros_tensor = torch.zeros(3, 4)
print("\nZeros Tensor:\n", zeros_tensor)

ones_tensor = torch.ones(2, 2)
print("\nOnes Tensor:\n", ones_tensor)

Empty Tensor:
 tensor([[6.1963e-04, 6.7333e+22, 2.7622e+29],
        [4.3219e+27, 4.3055e+21, 1.8060e+28]])

Zeros Tensor:
 tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

Ones Tensor:
 tensor([[1., 1.],
        [1., 1.]])


**Reasoning**:
Continuing with instruction 2c, the next step is to create random tensors using `torch.rand()` for uniform distribution and `torch.randn()` for normal distribution with specified shapes.



In [7]:
random_uniform_tensor = torch.rand(2, 3)
print("Random Uniform Tensor (torch.rand):
", random_uniform_tensor)

random_normal_tensor = torch.randn(3, 2)
print("\nRandom Normal Tensor (torch.randn):
", random_normal_tensor)

SyntaxError: unterminated string literal (detected at line 2) (ipython-input-2131514353.py, line 2)

**Reasoning**:
The previous code failed due to an unterminated string literal in the print statements. I will fix the print statements by correctly closing the string literals and adding the newline character.



In [8]:
random_uniform_tensor = torch.rand(2, 3)
print("Random Uniform Tensor (torch.rand):\n", random_uniform_tensor)

random_normal_tensor = torch.randn(3, 2)
print("\nRandom Normal Tensor (torch.randn):\n", random_normal_tensor)

Random Uniform Tensor (torch.rand):
 tensor([[0.7470, 0.2680, 0.5695],
        [0.6122, 0.2704, 0.4371]])

Random Normal Tensor (torch.randn):
 tensor([[-0.2243, -0.6883],
        [-0.6111, -0.5849],
        [ 0.2527,  0.0070]])


**Reasoning**:
Following instruction 3, the next step is to explore key tensor attributes such as `shape`, `dtype`, and `device` for some of the tensors created previously.



In [9]:
print("Attributes of tensor_from_list:")
print("Shape:", tensor_from_list.shape)
print("Data type:", tensor_from_list.dtype)
print("Device:", tensor_from_list.device)

print("\nAttributes of random_uniform_tensor:")
print("Shape:", random_uniform_tensor.shape)
print("Data type:", random_uniform_tensor.dtype)
print("Device:", random_uniform_tensor.device)

# Check if CUDA is available and print device info
if torch.cuda.is_available():
    print("\nCUDA is available. Tensor will be moved to GPU.")
    cuda_device = torch.device("cuda")
    gpu_tensor = tensor_from_list.to(cuda_device)
    print("Device of GPU tensor:", gpu_tensor.device)
else:
    print("\nCUDA is not available. Tensors are on CPU.")

Attributes of tensor_from_list:
Shape: torch.Size([2, 3])
Data type: torch.int64
Device: cpu

Attributes of random_uniform_tensor:
Shape: torch.Size([2, 3])
Data type: torch.float32
Device: cpu

CUDA is available. Tensor will be moved to GPU.
Device of GPU tensor: cuda:0


**Reasoning**:
Following instruction 4a, the next step is to demonstrate basic tensor manipulations by reshaping tensors using `.view()` and `.reshape()`.



In [10]:
original_tensor = torch.tensor([[1, 2, 3], [4, 5, 6]])
print("Original Tensor:\n", original_tensor)

# Reshape using .view()
reshaped_view_tensor = original_tensor.view(3, 2)
print("\nReshaped Tensor (view - 3x2):\n", reshaped_view_tensor)

# Reshape using .reshape()
reshaped_reshape_tensor = original_tensor.reshape(1, 6)
print("\nReshaped Tensor (reshape - 1x6):\n", reshaped_reshape_tensor)

# Reshape using -1 for automatic size calculation
reshaped_auto_tensor = original_tensor.reshape(-1, 3) # -1 means infer the size
print("\nReshaped Tensor (reshape - auto size):\n", reshaped_auto_tensor)

Original Tensor:
 tensor([[1, 2, 3],
        [4, 5, 6]])

Reshaped Tensor (view - 3x2):
 tensor([[1, 2],
        [3, 4],
        [5, 6]])

Reshaped Tensor (reshape - 1x6):
 tensor([[1, 2, 3, 4, 5, 6]])

Reshaped Tensor (reshape - auto size):
 tensor([[1, 2, 3],
        [4, 5, 6]])


**Reasoning**:
Following instruction 4b, the next step is to perform slicing and indexing operations on tensors.



In [11]:
my_tensor = torch.tensor([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])
print("Original Tensor:\n", my_tensor)

# Access a single element
print("\nSingle element at [0, 1]:", my_tensor[0, 1].item())

# Slice a row
print("\nFirst row:\n", my_tensor[0, :])

# Slice a column
print("\nSecond column:\n", my_tensor[:, 1])

# Slice a sub-tensor
print("\nSub-tensor (rows 0-1, cols 1-2):\n", my_tensor[0:2, 1:3])

# Using advanced indexing (list of indices)
print("\nElements at specific indices ([0,0], [1,2], [2,3]):\n", my_tensor[[0, 1, 2], [0, 2, 3]])

Original Tensor:
 tensor([[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]])

Single element at [0, 1]: 2

First row:
 tensor([1, 2, 3, 4])

Second column:
 tensor([ 2,  6, 10])

Sub-tensor (rows 0-1, cols 1-2):
 tensor([[2, 3],
        [6, 7]])

Elements at specific indices ([0,0], [1,2], [2,3]):
 tensor([ 1,  7, 12])


**Reasoning**:
Following instruction 5a, the next step is to demonstrate element-wise arithmetic operations (addition, subtraction, multiplication, division) on tensors.



In [12]:
tensor_a = torch.tensor([[1, 2], [3, 4]])
tensor_b = torch.tensor([[5, 6], [7, 8]])

print("Tensor A:\n", tensor_a)
print("Tensor B:\n", tensor_b)

# Addition
add_result = tensor_a + tensor_b
print("\nAddition (A + B):\n", add_result)

# Subtraction
sub_result = tensor_a - tensor_b
print("\nSubtraction (A - B):\n", sub_result)

# Multiplication (element-wise)
mul_result = tensor_a * tensor_b
print("\nElement-wise Multiplication (A * B):\n", mul_result)

# Division (element-wise)
div_result = tensor_a / tensor_b
print("\nElement-wise Division (A / B):\n", div_result)

# Scalar multiplication
scalar_mul_result = tensor_a * 2
print("\nScalar Multiplication (A * 2):\n", scalar_mul_result)

Tensor A:
 tensor([[1, 2],
        [3, 4]])
Tensor B:
 tensor([[5, 6],
        [7, 8]])

Addition (A + B):
 tensor([[ 6,  8],
        [10, 12]])

Subtraction (A - B):
 tensor([[-4, -4],
        [-4, -4]])

Element-wise Multiplication (A * B):
 tensor([[ 5, 12],
        [21, 32]])

Element-wise Division (A / B):
 tensor([[0.2000, 0.3333],
        [0.4286, 0.5000]])

Scalar Multiplication (A * 2):
 tensor([[2, 4],
        [6, 8]])


**Reasoning**:
Following instruction 5b, the next step is to demonstrate matrix multiplication using `torch.matmul()` or the `@` operator.



In [13]:
tensor_x = torch.tensor([[1, 2], [3, 4]])
tensor_y = torch.tensor([[5, 6], [7, 8]])

print("Tensor X:\n", tensor_x)
print("Tensor Y:\n", tensor_y)

# Matrix multiplication using torch.matmul()
matmul_result = torch.matmul(tensor_x, tensor_y)
print("\nMatrix Multiplication (torch.matmul):\n", matmul_result)

# Matrix multiplication using @ operator
operator_matmul_result = tensor_x @ tensor_y
print("\nMatrix Multiplication (@ operator):\n", operator_matmul_result)


Tensor X:
 tensor([[1, 2],
        [3, 4]])
Tensor Y:
 tensor([[5, 6],
        [7, 8]])

Matrix Multiplication (torch.matmul):
 tensor([[19, 22],
        [43, 50]])

Matrix Multiplication (@ operator):
 tensor([[19, 22],
        [43, 50]])


**Reasoning**:
Following instruction 5c, the next step is to demonstrate transposing a tensor using `.T` or `.transpose()`.



In [14]:
original_tensor_t = torch.tensor([[1, 2, 3], [4, 5, 6]])
print("Original Tensor:\n", original_tensor_t)

# Transpose using .T
transposed_t = original_tensor_t.T
print("\nTransposed Tensor (.T):\n", transposed_t)

# Transpose using .transpose()
transposed_transpose = original_tensor_t.transpose(0, 1)
print("\nTransposed Tensor (.transpose(0, 1)):\n", transposed_transpose)

Original Tensor:
 tensor([[1, 2, 3],
        [4, 5, 6]])

Transposed Tensor (.T):
 tensor([[1, 4],
        [2, 5],
        [3, 6]])

Transposed Tensor (.transpose(0, 1)):
 tensor([[1, 4],
        [2, 5],
        [3, 6]])


**Reasoning**:
Following instruction 6, the next step is to demonstrate moving a tensor to a different device (specifically, the GPU if available) using the `.to()` method, as already partially demonstrated in instruction 3.



In [15]:
tensor_on_cpu = torch.tensor([[1, 2], [3, 4]])
print("Tensor on CPU device:", tensor_on_cpu.device)

if torch.cuda.is_available():
    # Move tensor to GPU
    tensor_on_gpu = tensor_on_cpu.to("cuda")
    print("Tensor moved to GPU device:", tensor_on_gpu.device)
else:
    print("CUDA is not available. Tensor remains on CPU.")

Tensor on CPU device: cpu
Tensor moved to GPU device: cuda:0


## Automatic Differentiation with Autograd

### Subtask:
Learn about PyTorch's `autograd` engine, which enables automatic computation of gradients. We will explore how it works and its importance in neural network training.


**Reasoning**:
As per instruction 1, I will start by creating a tensor `x` and setting `requires_grad=True`. This tells PyTorch to track all operations on this tensor for automatic gradient computation.

**Reasoning**:
As per instruction 1, I will create a tensor `x` with `requires_grad=True` to signal PyTorch to track operations on it, and print the tensor to confirm its creation.



In [16]:
x = torch.tensor([1.0, 2.0, 3.0], requires_grad=True)
print("Tensor x with requires_grad=True:", x)

Tensor x with requires_grad=True: tensor([1., 2., 3.], requires_grad=True)


**Reasoning**:
Following instruction 2, I will perform a series of operations on `x` to create a computational graph by defining `y = x**2` and `z = y.sum()`.



In [17]:
y = x**2
z = y.sum()
print("Tensor y (x**2):", y)
print("Tensor z (y.sum()):", z)

Tensor y (x**2): tensor([1., 4., 9.], grad_fn=<PowBackward0>)
Tensor z (y.sum()): tensor(14., grad_fn=<SumBackward0>)


**Reasoning**:
Now that the computational graph is set up (`z = (x**2).sum()`), as per instruction 3, I will compute the gradients of `z` with respect to `x` using the `.backward()` method on `z`.

**Reasoning**:
Now that the computational graph is set up (`z = (x**2).sum()`), as per instruction 3, I will compute the gradients of `z` with respect to `x` using the `.backward()` method on `z`.



In [18]:
z.backward()
print("Gradients computed successfully.")

Gradients computed successfully.


**Reasoning**:
As per instruction 4, I will access the gradients of `x` using `x.grad` and print the result to inspect the computed gradients.

**Reasoning**:
As per instruction 4, I will now access the computed gradients of `x` using `x.grad` and print them to verify the result.



In [19]:
print("Gradients of x (x.grad):", x.grad)

Gradients of x (x.grad): tensor([2., 4., 6.])


**Reasoning**:
As per instruction 5, I will now demonstrate how gradients accumulate. I will define a new operation `z2 = (x * 3).sum()`, perform a backward pass on `z2`, and then print `x.grad` again to observe the accumulation of gradients.

**Reasoning**:
As per instruction 5, I will now define `z2 = (x * 3).sum()`, then call `z2.backward()` to accumulate gradients, and finally print `x.grad` to observe the accumulated gradients.



In [20]:
z2 = (x * 3).sum()
print("Tensor z2 ((x * 3).sum()):", z2)
z2.backward()
print("Gradients of x after accumulating (x.grad):", x.grad)

Tensor z2 ((x * 3).sum()): tensor(18., grad_fn=<SumBackward0>)
Gradients of x after accumulating (x.grad): tensor([5., 7., 9.])


**Reasoning**:
As per instruction 6, I will now reset the gradients of `x` using `x.grad.zero_()` and then confirm that they are indeed reset to zero by printing `x.grad`.

**Reasoning**:
As per instruction 6, I will now reset the gradients of `x` using `x.grad.zero_()` and then confirm that they are indeed reset to zero by printing `x.grad`.



In [21]:
x.grad.zero_()
print("Gradients of x after reset (x.grad):", x.grad)

Gradients of x after reset (x.grad): tensor([0., 0., 0.])


**Reasoning**:
As per instruction 7, I will now illustrate disabling gradient tracking using `with torch.no_grad():` block and demonstrate detaching a tensor from the computation graph using `.detach()`.

**Reasoning**:
As per instruction 7, I will now demonstrate disabling gradient tracking using `with torch.no_grad():` and detaching a tensor using `.detach()` and print the results to show their effect.



In [22]:
x_no_grad = torch.tensor([4.0, 5.0, 6.0], requires_grad=True)
print("Original tensor for no_grad demo:", x_no_grad)

# Disabling gradient tracking with torch.no_grad()
with torch.no_grad():
    y_no_grad = x_no_grad * 2
    print("Tensor inside no_grad block (y_no_grad):
", y_no_grad)
    print("y_no_grad.requires_grad:", y_no_grad.requires_grad)

# Detaching a tensor
x_detach = torch.tensor([7.0, 8.0, 9.0], requires_grad=True)
y_detach = x_detach * 2
print("\nOriginal tensor for detach demo:", x_detach)
print("Tensor y_detach (x_detach * 2):
", y_detach)
print("y_detach.requires_grad:", y_detach.requires_grad)

detached_tensor = y_detach.detach()
print("Detached tensor (detached_tensor):
", detached_tensor)
print("detached_tensor.requires_grad:", detached_tensor.requires_grad)


SyntaxError: unterminated string literal (detected at line 7) (ipython-input-3449167404.py, line 7)

**Reasoning**:
The previous code failed due to an unterminated string literal in multiple print statements. I will fix the print statements by correctly closing the string literals and ensuring that newline characters are correctly placed within the string or concatenated.



In [23]:
x_no_grad = torch.tensor([4.0, 5.0, 6.0], requires_grad=True)
print("Original tensor for no_grad demo:", x_no_grad)

# Disabling gradient tracking with torch.no_grad()
with torch.no_grad():
    y_no_grad = x_no_grad * 2
    print("Tensor inside no_grad block (y_no_grad):\n", y_no_grad)
    print("y_no_grad.requires_grad:", y_no_grad.requires_grad)

# Detaching a tensor
x_detach = torch.tensor([7.0, 8.0, 9.0], requires_grad=True)
y_detach = x_detach * 2
print("\nOriginal tensor for detach demo:", x_detach)
print("Tensor y_detach (x_detach * 2):\n", y_detach)
print("y_detach.requires_grad:", y_detach.requires_grad)

detached_tensor = y_detach.detach()
print("Detached tensor (detached_tensor):\n", detached_tensor)
print("detached_tensor.requires_grad:", detached_tensor.requires_grad)

Original tensor for no_grad demo: tensor([4., 5., 6.], requires_grad=True)
Tensor inside no_grad block (y_no_grad):
 tensor([ 8., 10., 12.])
y_no_grad.requires_grad: False

Original tensor for detach demo: tensor([7., 8., 9.], requires_grad=True)
Tensor y_detach (x_detach * 2):
 tensor([14., 16., 18.], grad_fn=<MulBackward0>)
y_detach.requires_grad: True
Detached tensor (detached_tensor):
 tensor([14., 16., 18.])
detached_tensor.requires_grad: False


## Building Neural Networks with nn.Module

### Subtask:
Dive into `torch.nn` to define and build neural network architectures. This step will cover layers, activation functions, and assembling simple models.


**Reasoning**:
As per instruction 1, I will now define a simple neural network class using `torch.nn.Module`. This class will encapsulate a single linear layer, and I will implement the `__init__` method to define the layer and the `forward` method to specify the computation flow.

**Reasoning**:
As per instruction 1, I will now define the `SimpleNN` class using `torch.nn.Module` with a single linear layer, implementing both the `__init__` and `forward` methods.



In [24]:
import torch.nn as nn

class SimpleNN(nn.Module):
    def __init__(self, input_size, output_size):
        super(SimpleNN, self).__init__()
        self.linear = nn.Linear(input_size, output_size)

    def forward(self, x):
        return self.linear(x)

print("SimpleNN class defined successfully.")

SimpleNN class defined successfully.


**Reasoning**:
As per instruction 2, I will now instantiate the `SimpleNN` model by creating an object of the class, specifying arbitrary `input_size` and `output_size` for demonstration.

**Reasoning**:
Following the reasoning in the previous markdown cell, I will now instantiate the `SimpleNN` model with a chosen `input_size` and `output_size` to create a working model instance.



In [25]:
input_size = 10
output_size = 1
model = SimpleNN(input_size, output_size)
print("SimpleNN model instantiated successfully:")
print(model)

SimpleNN model instantiated successfully:
SimpleNN(
  (linear): Linear(in_features=10, out_features=1, bias=True)
)


**Reasoning**:
As per instruction 3, the model architecture has already been printed in the previous step, confirming the layers and parameters within the network. This completes instruction 3.

**Reasoning**:
Following instruction 4, I will now demonstrate a forward pass through the instantiated model. This involves creating a dummy input tensor with the correct input shape and passing it through the model, then printing the shape of the output tensor to confirm the network's output.

**Reasoning**:
As per instruction 4, I will now create a dummy input tensor with the appropriate input size and pass it through the `model` to perform a forward pass. I will then print the shape of the resulting output tensor to confirm the network's output.



In [26]:
import torch

# Create a dummy input tensor (e.g., a batch of 1 with input_size features)
dummy_input = torch.randn(1, input_size)
print("Dummy input tensor shape:", dummy_input.shape)

# Perform a forward pass
output = model(dummy_input)

print("\nOutput tensor from forward pass:", output)
print("Output tensor shape:", output.shape)

Dummy input tensor shape: torch.Size([1, 10])

Output tensor from forward pass: tensor([[-0.9817]], grad_fn=<AddmmBackward0>)
Output tensor shape: torch.Size([1, 1])


## Defining Loss Functions and Optimizers

### Subtask:
Understand different types of loss functions (e.g., CrossEntropyLoss, MSELoss) and optimization algorithms (e.g., SGD, Adam) used to train neural networks.


**Reasoning**:
As per instruction 1, I will import the necessary modules, `torch.nn` for loss functions and `torch.optim` for optimizers.

**Reasoning**:
Following the previous markdown explanation, I will now import the `torch.nn` and `torch.optim` modules as instructed, which are essential for defining loss functions and optimizers.



In [27]:
import torch.nn as nn
import torch.optim as optim

print("torch.nn and torch.optim modules imported successfully.")

torch.nn and torch.optim modules imported successfully.


**Reasoning**:
As per instruction 2, I will now instantiate a Mean Squared Error (MSE) loss function. This loss is commonly used for regression tasks.

**Reasoning**:
Following the previous markdown explanation and instruction 2, I will now instantiate the `nn.MSELoss` function, which is suitable for regression problems.



In [28]:
mse_loss = nn.MSELoss()
print("Mean Squared Error (MSE) loss function instantiated successfully.")

Mean Squared Error (MSE) loss function instantiated successfully.


**Reasoning**:
As per instruction 3, I will now instantiate a Cross Entropy Loss function. This loss is typically used for classification tasks.

**Reasoning**:
Following the previous markdown explanation and instruction 3, I will now instantiate the `nn.CrossEntropyLoss` function, which is suitable for classification problems.



In [29]:
cross_entropy_loss = nn.CrossEntropyLoss()
print("Cross Entropy Loss function instantiated successfully.")

Cross Entropy Loss function instantiated successfully.


**Reasoning**:
As per instruction 4, I will now instantiate an SGD (Stochastic Gradient Descent) optimizer. This optimizer is commonly used for training neural networks and requires the model's parameters and a learning rate.

**Reasoning**:
Following the previous markdown explanation and instruction 4, I will now instantiate the `optim.SGD` optimizer, passing the previously defined `model.parameters()` and a learning rate of 0.01.



In [30]:
sgd_optimizer = optim.SGD(model.parameters(), lr=0.01)
print("SGD optimizer instantiated successfully.")

SGD optimizer instantiated successfully.


**Reasoning**:
As per instruction 5, I will now instantiate an Adam optimizer. This optimizer is also commonly used for training neural networks and requires the model's parameters and a learning rate.

**Reasoning**:
Following the previous markdown explanation and instruction 5, I will now instantiate the `optim.Adam` optimizer, passing the previously defined `model.parameters()` and a learning rate of 0.001.



In [31]:
adam_optimizer = optim.Adam(model.parameters(), lr=0.001)
print("Adam optimizer instantiated successfully.")

Adam optimizer instantiated successfully.


## Data Loading and Preprocessing

### Subtask:
Learn how to prepare data for training using `torch.utils.data.Dataset` and `DataLoader` for efficient batch processing.


**Reasoning**:
As per instruction 1, I will start by importing the necessary `Dataset` and `DataLoader` classes from `torch.utils.data`.

**Reasoning**:
As per instruction 1, I will start by importing the necessary `Dataset` and `DataLoader` classes from `torch.utils.data`.



In [32]:
from torch.utils.data import Dataset, DataLoader

print("Dataset and DataLoader imported successfully.")

Dataset and DataLoader imported successfully.


**Reasoning**:
As per instruction 2, I will now define a custom `MyDataset` class that inherits from `torch.utils.data.Dataset`. This class will include methods for initialization, length, and item retrieval to encapsulate dummy features and labels.

**Reasoning**:
As per instruction 2, I will now define the `MyDataset` class with `__init__`, `__len__`, and `__getitem__` methods to handle dummy features and labels.



In [34]:
import torch

class MyDataset(Dataset):
    def __init__(self, num_samples, num_features, num_labels=1):
        self.features = torch.randn(num_samples, num_features)
        self.labels = torch.randint(0, 2, (num_samples, num_labels)).float() # Binary labels for demonstration

    def __len__(self):
        return len(self.features)

    def __getitem__(self, idx):
        return self.features[idx], self.labels[idx]

print("MyDataset class defined successfully.")

MyDataset class defined successfully.


**Reasoning**:
As per instruction 3, I will now instantiate the `MyDataset` with specified `num_samples`, `num_features`, and `num_labels`.

**Reasoning**:
Following the previous markdown explanation, I will now instantiate the `MyDataset` class with `num_samples=10`, `num_features=5`, and `num_labels=1` and print its length to verify its creation.



In [35]:
num_samples = 10
num_features = 5
num_labels = 1

my_custom_dataset = MyDataset(num_samples, num_features, num_labels)
print(f"MyDataset instantiated successfully with {len(my_custom_dataset)} samples.")
print(f"First feature sample shape: {my_custom_dataset[0][0].shape}")
print(f"First label sample shape: {my_custom_dataset[0][1].shape}")

MyDataset instantiated successfully with 10 samples.
First feature sample shape: torch.Size([5])
First label sample shape: torch.Size([1])


**Reasoning**:
As per instruction 4, I will now create a `DataLoader` instance, passing the `my_custom_dataset`, a `batch_size` of 2, and `shuffle=True` to enable batch processing and shuffling.

**Reasoning**:
Following the previous markdown explanation and instruction 4, I will now create a `DataLoader` instance using the `my_custom_dataset`, a `batch_size` of 2, and `shuffle=True` to prepare data for efficient batch processing.



In [36]:
batch_size = 2
my_dataloader = DataLoader(my_custom_dataset, batch_size=batch_size, shuffle=True)
print(f"DataLoader instantiated successfully with batch size {batch_size}.")

DataLoader instantiated successfully with batch size 2.


**Reasoning**:
As per instruction 5, I will now iterate through the `DataLoader` using a `for` loop, printing the shape of the features and labels for each batch to demonstrate batching.

**Reasoning**:
Following the previous markdown explanation and instruction 5, I will now iterate through the `my_dataloader` using a `for` loop to print the shape of the features and labels for each batch, demonstrating how data is batched.



In [37]:
for batch_idx, (features, labels) in enumerate(my_dataloader):
    print(f"Batch {batch_idx+1}:")
    print(f"  Features shape: {features.shape}")
    print(f"  Labels shape: {labels.shape}")

Batch 1:
  Features shape: torch.Size([2, 5])
  Labels shape: torch.Size([2, 1])
Batch 2:
  Features shape: torch.Size([2, 5])
  Labels shape: torch.Size([2, 1])
Batch 3:
  Features shape: torch.Size([2, 5])
  Labels shape: torch.Size([2, 1])
Batch 4:
  Features shape: torch.Size([2, 5])
  Labels shape: torch.Size([2, 1])
Batch 5:
  Features shape: torch.Size([2, 5])
  Labels shape: torch.Size([2, 1])


## Implementing a Training Loop

### Subtask:
Construct a complete training loop, integrating model definition, data loading, loss calculation, backpropagation, and optimizer steps.


**Reasoning**:
As per instruction 1, I will define a more complex neural network class, `NeuralNetwork`, that includes at least two linear layers and an activation function (`nn.ReLU`), and define its `__init__` and `forward` methods.



In [39]:
import torch.nn as nn

class NeuralNetwork(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(NeuralNetwork, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        out = self.fc1(x)
        out = self.relu(out)
        out = self.fc2(out)
        return out

print("NeuralNetwork class defined successfully.")

NeuralNetwork class defined successfully.


**Reasoning**:
As per instruction 2, I will now instantiate the `NeuralNetwork` model, using `num_features` as the input size, `output_size` as 1, and a chosen `hidden_size`.



In [40]:
hidden_size = 64
model_nn = NeuralNetwork(num_features, hidden_size, num_labels)
print("NeuralNetwork model instantiated successfully:")
print(model_nn)

NeuralNetwork model instantiated successfully:
NeuralNetwork(
  (fc1): Linear(in_features=5, out_features=64, bias=True)
  (relu): ReLU()
  (fc2): Linear(in_features=64, out_features=1, bias=True)
)


**Reasoning**:
As per instruction 3, I will now instantiate a suitable loss function for binary classification, `nn.BCEWithLogitsLoss()`, which is commonly used for this type of task as it combines a sigmoid activation and binary cross-entropy loss for numerical stability.



In [41]:
loss_function = nn.BCEWithLogitsLoss()
print("nn.BCEWithLogitsLoss instantiated successfully.")

nn.BCEWithLogitsLoss instantiated successfully.


**Reasoning**:
As per instruction 4, I will now instantiate an Adam optimizer, passing the `model_nn`'s parameters and a learning rate of 0.001.



In [42]:
optimizer = optim.Adam(model_nn.parameters(), lr=0.001)
print("Adam optimizer instantiated successfully.")

Adam optimizer instantiated successfully.


**Reasoning**:
As per instruction 5, I will now implement the complete training loop. This involves iterating through epochs, processing data in batches, performing forward passes, calculating loss, zeroing gradients, backpropagating, and updating model parameters, finally printing the loss for each epoch.



In [43]:
num_epochs = 10

print("Starting training loop...")

for epoch in range(num_epochs):
    total_loss = 0
    for batch_idx, (features, labels) in enumerate(my_dataloader):
        # Forward pass
        outputs = model_nn(features)
        loss = loss_function(outputs, labels)

        # Backward and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        total_loss += loss.item()

    # Print loss for every epoch
    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {total_loss / len(my_dataloader):.4f}')

print("Training loop completed.")

Starting training loop...
Epoch [1/10], Loss: 0.7053
Epoch [2/10], Loss: 0.6948
Epoch [3/10], Loss: 0.6875
Epoch [4/10], Loss: 0.6781
Epoch [5/10], Loss: 0.6712
Epoch [6/10], Loss: 0.6642
Epoch [7/10], Loss: 0.6571
Epoch [8/10], Loss: 0.6523
Epoch [9/10], Loss: 0.6449
Epoch [10/10], Loss: 0.6397
Training loop completed.


## Saving and Loading Models

### Subtask:
Learn how to save and load trained PyTorch models, allowing for model persistence and deployment.


**Reasoning**:
As per instruction 1, I will save the state dictionary of the trained model (`model_nn`) to a file named 'model_state_dict.pth' using `torch.save()`.



In [44]:
torch.save(model_nn.state_dict(), 'model_state_dict.pth')
print("Model state dictionary saved to 'model_state_dict.pth'.")

Model state dictionary saved to 'model_state_dict.pth'.


**Reasoning**:
As per instruction 2, I will create a new instance of the `NeuralNetwork` class (`loaded_model`) with the same architecture as the original model.



In [45]:
loaded_model = NeuralNetwork(num_features, hidden_size, num_labels)
print("New NeuralNetwork instance ('loaded_model') created successfully:")
print(loaded_model)

New NeuralNetwork instance ('loaded_model') created successfully:
NeuralNetwork(
  (fc1): Linear(in_features=5, out_features=64, bias=True)
  (relu): ReLU()
  (fc2): Linear(in_features=64, out_features=1, bias=True)
)


**Reasoning**:
As per instruction 3, I will now load the saved state dictionary from 'model_state_dict.pth' into the `loaded_model` using `torch.load()` and `loaded_model.load_state_dict()`.



In [46]:
loaded_model.load_state_dict(torch.load('model_state_dict.pth'))
print("State dictionary loaded into 'loaded_model' successfully.")

State dictionary loaded into 'loaded_model' successfully.


**Reasoning**:
As per instruction 4, I will now set the `loaded_model` to evaluation mode using `loaded_model.eval()` to disable dropout and batch normalization for inference.



In [47]:
loaded_model.eval()
print("Loaded model set to evaluation mode.")

Loaded model set to evaluation mode.


**Reasoning**:
As per instruction 5, I will now create a dummy input tensor with a shape of `(1, num_features)` using `torch.randn()` to be used for forward passes.



In [49]:
dummy_input = torch.randn(1, num_features)
print("Dummy input tensor created successfully with shape:", dummy_input.shape)

Dummy input tensor created successfully with shape: torch.Size([1, 5])


**Reasoning**:
As per instruction 6 and 7, I will now perform a forward pass using both the original `model_nn` and the `loaded_model` with the `dummy_input`, and then print their outputs to verify that they produce similar results.



In [50]:
model_nn.eval()
with torch.no_grad():
    original_model_output = model_nn(dummy_input)
    loaded_model_output = loaded_model(dummy_input)

print("Output from original model:", original_model_output)
print("Output from loaded model:", loaded_model_output)

# Verify if outputs are similar (e.g., using torch.allclose)
if torch.allclose(original_model_output, loaded_model_output):
    print("Outputs from original and loaded models are similar.")
else:
    print("Outputs from original and loaded models are NOT similar.")

Output from original model: tensor([[0.4020]])
Output from loaded model: tensor([[0.4020]])
Outputs from original and loaded models are similar.


## Practical Example: Image Classification

### Subtask:
Apply all learned concepts to a hands-on project, such as building and training a basic image classifier on a public dataset like MNIST or FashionMNIST.


**Reasoning**:
As per instruction 1, I will start by importing all the necessary libraries for building and training an image classifier: `torch`, `torchvision`, `torch.nn`, and `torch.optim`.



In [51]:
import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import torch.optim as optim

print("Necessary libraries imported successfully.")

Necessary libraries imported successfully.


**Reasoning**:
As per instruction 2, I will now define data transformations for the training and test datasets. This includes converting images to tensors and normalizing pixel values for both sets.



In [52]:
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5,), (0.5,)) # Normalize with mean 0.5 and std 0.5
])

print("Data transformations defined successfully.")

Data transformations defined successfully.


**Reasoning**:
As per instruction 3, I will now load the FashionMNIST training and test datasets using `torchvision.datasets.FashionMNIST`. This involves specifying `train=True` for the training set and `train=False` for the test set, applying the defined `transform`, and setting `download=True` to ensure the datasets are available.



In [53]:
train_dataset = torchvision.datasets.FashionMNIST(root='./data', train=True, download=True, transform=transform)
test_dataset = torchvision.datasets.FashionMNIST(root='./data', train=False, download=True, transform=transform)

print(f"FashionMNIST training dataset loaded with {len(train_dataset)} samples.")
print(f"FashionMNIST test dataset loaded with {len(test_dataset)} samples.")

100%|██████████| 26.4M/26.4M [00:02<00:00, 12.5MB/s]
100%|██████████| 29.5k/29.5k [00:00<00:00, 211kB/s]
100%|██████████| 4.42M/4.42M [00:01<00:00, 3.94MB/s]
100%|██████████| 5.15k/5.15k [00:00<00:00, 8.74MB/s]

FashionMNIST training dataset loaded with 60000 samples.
FashionMNIST test dataset loaded with 10000 samples.


**Reasoning**:
As per instruction 4, I will now create `DataLoader` instances for both the training and test sets, specifying a `batch_size` and enabling shuffling for the training data.



In [55]:
from torch.utils.data import DataLoader

batch_size = 64

train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

print(f"Training DataLoader created with batch size {batch_size}.")
print(f"Test DataLoader created with batch size {batch_size}.")

Training DataLoader created with batch size 64.
Test DataLoader created with batch size 64.


**Reasoning**:
As per instruction 5, I will now define a Convolutional Neural Network (CNN) class, inheriting from `nn.Module`. This class will include at least two convolutional layers, activation functions, pooling layers, fully connected layers, and a `forward` method that handles the flattening of convolutional outputs.



In [56]:
class CNN(nn.Module):
    def __init__(self):
        super(CNN, self).__init__()
        # First convolutional block
        self.conv1 = nn.Conv2d(in_channels=1, out_channels=32, kernel_size=3, padding=1)
        self.relu1 = nn.ReLU()
        self.pool1 = nn.MaxPool2d(kernel_size=2, stride=2)

        # Second convolutional block
        self.conv2 = nn.Conv2d(in_channels=32, out_channels=64, kernel_size=3, padding=1)
        self.relu2 = nn.ReLU()
        self.pool2 = nn.MaxPool2d(kernel_size=2, stride=2)

        # Fully connected layers
        # The input size to the first FC layer depends on the output size of the last pooling layer.
        # For a 28x28 input image: (28/2)/2 = 7
        # So, it's 64 channels * 7 * 7 spatial dimensions
        self.fc1 = nn.Linear(64 * 7 * 7, 128)
        self.relu3 = nn.ReLU()
        self.fc2 = nn.Linear(128, 10) # 10 classes for FashionMNIST

    def forward(self, x):
        # Conv -> ReLU -> Pool 1
        out = self.conv1(x)
        out = self.relu1(out)
        out = self.pool1(out)

        # Conv -> ReLU -> Pool 2
        out = self.conv2(out)
        out = self.relu2(out)
        out = self.pool2(out)

        # Flatten the output for the fully connected layer
        out = out.view(out.size(0), -1) # Flatten all dimensions except batch size

        # Fully connected layers
        out = self.fc1(out)
        out = self.relu3(out)
        out = self.fc2(out)
        return out

print("CNN class defined successfully.")

CNN class defined successfully.


**Reasoning**:
As per instruction 6, I will now instantiate the CNN model and move it to a GPU if available. This involves creating an instance of the `CNN` class and checking for CUDA availability to set the device.



In [57]:
model_cnn = CNN()
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model_cnn.to(device)

print(f"CNN model instantiated and moved to {device} successfully.")
print(model_cnn)

CNN model instantiated and moved to cuda successfully.
CNN(
  (conv1): Conv2d(1, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (relu1): ReLU()
  (pool1): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (conv2): Conv2d(32, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (relu2): ReLU()
  (pool2): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (fc1): Linear(in_features=3136, out_features=128, bias=True)
  (relu3): ReLU()
  (fc2): Linear(in_features=128, out_features=10, bias=True)
)


**Reasoning**:
As per instruction 7, I will now define the loss function, `nn.CrossEntropyLoss`, suitable for multi-class classification, and an `optim.Adam` optimizer, passing the `model_cnn`'s parameters and a learning rate.



In [58]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model_cnn.parameters(), lr=0.001)

print("Loss function (CrossEntropyLoss) and optimizer (Adam) defined successfully.")

Loss function (CrossEntropyLoss) and optimizer (Adam) defined successfully.


**Reasoning**:
As per instruction 8, I will now implement the complete training loop. This involves iterating through epochs, processing data in batches, moving data to the device, performing forward passes, calculating loss, zeroing gradients, backpropagating, and updating model parameters, finally printing the loss for each epoch.



In [59]:
num_epochs = 5

print("Starting CNN training loop...")

for epoch in range(num_epochs):
    model_cnn.train() # Set model to training mode
    running_loss = 0.0
    for i, (images, labels) in enumerate(train_loader):
        images, labels = images.to(device), labels.to(device)

        # Forward pass
        outputs = model_cnn(images)
        loss = criterion(outputs, labels)

        # Backward and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        running_loss += loss.item()

    # Print loss for every epoch
    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {running_loss / len(train_loader):.4f}')

print("CNN training loop completed.")

Starting CNN training loop...
Epoch [1/5], Loss: 0.4299
Epoch [2/5], Loss: 0.2737
Epoch [3/5], Loss: 0.2281
Epoch [4/5], Loss: 0.1952
Epoch [5/5], Loss: 0.1704
CNN training loop completed.


**Reasoning**:
As per instruction 9, I will now implement an evaluation loop for the trained CNN model on the test set. This involves setting the model to evaluation mode, disabling gradient calculations, iterating through the test DataLoader, performing forward passes, calculating predictions, and finally computing and printing the test accuracy.



In [60]:
model_cnn.eval() # Set the model to evaluation mode

correct = 0
total = 0
with torch.no_grad(): # Disable gradient calculation for evaluation
    for images, labels in test_loader:
        images, labels = images.to(device), labels.to(device)

        outputs = model_cnn(images)
        _, predicted = torch.max(outputs.data, 1) # Get the index of the max log-probability

        total += labels.size(0)
        correct += (predicted == labels).sum().item()

accuracy = 100 * correct / total
print(f'Accuracy of the model on the {total} test images: {accuracy:.2f}%')


Accuracy of the model on the 10000 test images: 91.25%


## Final Task

### Subtask:
Summarize the key PyTorch concepts covered and suggest next steps for further learning and exploration in areas like CNNs, RNNs, and advanced architectures.


## Summary:

The task successfully demonstrated the creation, manipulation, and fundamental operations of Tensors using PyTorch, along with key concepts for building and training neural networks.

### Data Analysis Key Findings

*   **Tensor Fundamentals:** Tensors were successfully created from various sources (Python lists, NumPy arrays) and with specific initializations (`torch.empty`, `torch.zeros`, `torch.ones`, `torch.rand`, `torch.randn`). Their attributes (shape, data type, device) were inspected, and basic operations like reshaping, slicing, element-wise arithmetic, matrix multiplication, and transposition were performed.
*   **Automatic Differentiation (`autograd`):** The core mechanism of `autograd` was explored by setting `requires_grad=True` for a tensor, building a computational graph, and computing gradients using `backward()`. Gradient accumulation and resetting with `zero_()` were demonstrated, along with disabling gradient tracking using `torch.no_grad()` and `detach()`.
*   **Neural Network Construction (`nn.Module`):** A simple neural network was defined by subclassing `torch.nn.Module`, incorporating `Linear` layers, and implementing a `forward` method. The model's architecture was inspected, and a forward pass was successfully executed with dummy input.
*   **Loss Functions and Optimizers:** Essential components for training, `nn.MSELoss` (for regression) and `nn.CrossEntropyLoss` (for classification) were instantiated. Common optimizers, `optim.SGD` and `optim.Adam`, were also successfully set up.
*   **Data Handling (`Dataset` and `DataLoader`):** A custom `Dataset` class was created to encapsulate dummy features and labels. The `DataLoader` was then used to efficiently batch and shuffle this data, demonstrating batch iteration with correct tensor shapes.
*   **Training Loop Implementation:** A complete training loop was implemented, integrating a defined `NeuralNetwork` model, a `BCEWithLogitsLoss` loss function, and an `Adam` optimizer. Over 10 epochs, the training loss decreased from approximately 0.7053 to 0.6397, indicating the model was learning.
*   **Model Persistence:** The process of saving a model's state dictionary (`torch.save(model.state_dict(), ...)`) and loading it into a new model instance (`loaded_model.load_state_dict(torch.load(...))`) was successfully demonstrated. Outputs from the original and loaded models were verified to be identical.
*   **Practical Application (Image Classification):** All learned concepts were integrated into a practical image classification task on the FashionMNIST dataset. A Convolutional Neural Network (CNN) was built, trained for 5 epochs (with training loss decreasing from ~0.4299 to ~0.1704), and achieved a test accuracy of **91.25%**.

### Insights or Next Steps

*   **Deepen Understanding of `autograd`:** Explore more complex computational graphs, custom `autograd.Function` implementations, and interactions with control flow to gain a more advanced understanding of gradient computation.
*   **Explore Advanced Architectures:** Extend the practical application by implementing more complex CNN architectures (e.g., ResNet, VGG) or venturing into Recurrent Neural Networks (RNNs) for sequential data processing, and Transformer networks for state-of-the-art models.
